File: LiveKit/LiveSession.m
//
//  LiveSession.m
//  LiveKit
//
//  Created by LaiFeng on 16/5/20.
//  Copyright © 2016年 LaiFeng All rights reserved.
//

#import "LiveSession.h"
#import "AudioCapture.h"
#import "LFHardwareVideoEncoder.h"
#import "LFHardwareAudioEncoder.h"
#import "LFH264VideoEncoder.h"
#import "StreamRTMPSocket.h"
#import "LiveStreamInfo.h"
//#import "LFGPUImageBeautyFilter.h"
#import "LFH264VideoEncoder.h"

#import <LiveKit/LiveKit-Swift.h>

@interface LiveSession ()<AudioCaptureDelegate, VideoCaptureDelegate, AudioEncodingDelegate, VideoEncodingDelegate, StreamSocketDelegate>

/// 音频配置
@property (nonatomic, strong) LiveAudioConfiguration *audioConfiguration;
/// 视频配置
@property (nonatomic, strong) LiveVideoConfiguration *videoConfiguration;
/// 声音采集
@property (nonatomic, strong) AudioCapture *audioCaptureSource;
/// 视频采集
//@property (nonatomic, strong) VideoCapture *videoCaptureSource;

@property (nonatomic, strong) LiveVideoCapture *videoCaptureSource;

/// 音频编码
@property (nonatomic, strong) id<AudioEncoding> audioEncoder;
/// 视频编码
@property (nonatomic, strong) id<VideoEncoding> videoEncoder;
/// 上传
@property (nonatomic, strong) id<StreamSocket> socket;


#pragma mark -- 内部标识
/// 调试信息
@property (nonatomic, strong) LiveDebug *debugInfo;
/// 流信息
@property (nonatomic, strong) LiveStreamInfo *streamInfo;
/// 是否开始上传
@property (nonatomic, assign) BOOL uploading;
/// 当前状态
@property (nonatomic, assign, readwrite) LiveState state;
/// 当前直播type
@property (nonatomic, assign, readwrite) LiveCaptureTypeMask captureType;
/// 时间戳锁
@property (nonatomic, strong) dispatch_semaphore_t lock;


@end

/**  时间戳 */
#define NOW (CACurrentMediaTime()*1000)
#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@interface LiveSession ()

/// 上传相对时间戳
@property (nonatomic, assign) uint64_t relativeTimestamps;
/// 音视频是否对齐
@property (nonatomic, assign) BOOL AVAlignment;
/// 当前是否采集到了音频
@property (nonatomic, assign) BOOL hasCaptureAudio;
/// 当前是否采集到了关键帧
@property (nonatomic, assign) BOOL hasKeyFrameVideo;

@end

@implementation LiveSession

#pragma mark -- LifeCycle
- (instancetype)initWithAudioConfiguration:(nullable LiveAudioConfiguration *)audioConfiguration videoConfiguration:(nullable LiveVideoConfiguration *)videoConfiguration {
    return [self initWithAudioConfiguration:audioConfiguration videoConfiguration:videoConfiguration captureType:LiveCaptureDefaultMask];
}

- (nullable instancetype)initWithAudioConfiguration:(nullable LiveAudioConfiguration *)audioConfiguration videoConfiguration:(nullable LiveVideoConfiguration *)videoConfiguration captureType:(LiveCaptureTypeMask)captureType{
    if((captureType & LiveCaptureMaskAudio || captureType & LiveInputMaskAudio) && !audioConfiguration) @throw [NSException exceptionWithName:@"LiveSession init error" reason:@"audioConfiguration is nil " userInfo:nil];
    if((captureType & LiveCaptureMaskVideo || captureType & LiveInputMaskVideo) && !videoConfiguration) @throw [NSException exceptionWithName:@"LiveSession init error" reason:@"videoConfiguration is nil " userInfo:nil];
    if (self = [super init]) {
        _audioConfiguration = audioConfiguration;
        _videoConfiguration = videoConfiguration;
        _adaptiveBitrate = NO;
        _captureType = captureType;
    }
    return self;
}

- (void)dealloc {
    [_videoCaptureSource end];
    _audioCaptureSource.running = NO;
}

#pragma mark -- CustomMethod
- (void)startLive:(LiveStreamInfo *)streamInfo {
    if (!streamInfo) return;
    _streamInfo = streamInfo;
    _streamInfo.videoConfiguration = _videoConfiguration;
    _streamInfo.audioConfiguration = _audioConfiguration;
    [self.socket start];
}

- (void)stopLive {
    self.uploading = NO;
    [self.socket stop];
    self.socket = nil;
}

- (void)pushVideo:(nullable CVPixelBufferRef)pixelBuffer{
    if(self.captureType & LiveInputMaskVideo){
        if (self.uploading) [self.videoEncoder encodeVideoData:pixelBuffer timeStamp:NOW];
    }
}

- (void)pushAudio:(nullable NSData*)audioData{
    if(self.captureType & LiveInputMaskAudio){
        if (self.uploading) [self.audioEncoder encodeAudioData:audioData timeStamp:NOW];
    }
}

#pragma mark -- PrivateMethod
- (void)pushSendBuffer:(LFFrame*)frame{
    if(self.relativeTimestamps == 0){
        self.relativeTimestamps = frame.timestamp;
    }
    frame.timestamp = [self uploadTimestamp:frame.timestamp];
    [self.socket sendFrame:frame];
}

#pragma mark -- CaptureDelegate
- (void)captureOutput:(nullable AudioCapture *)capture audioData:(nullable NSData*)audioData {
    if (self.uploading) [self.audioEncoder encodeAudioData:audioData timeStamp:NOW];
}

- (void)captureOutputWithCapture:(LiveVideoCapture * _Nullable)capture pixelBuffer:(CVPixelBufferRef _Nullable)pixelBuffer {
    if (self.uploading) [self.videoEncoder encodeVideoData:pixelBuffer timeStamp:NOW];
}

#pragma mark -- EncoderDelegate
- (void)audioEncoder:(nullable id<AudioEncoding>)encoder audioFrame:(nullable AudioFrame *)frame {
    if (self.uploading){
        self.hasCaptureAudio = YES;
        if(self.AVAlignment) [self pushSendBuffer:frame];
    }
}

- (void)videoEncoder:(nullable id<VideoEncoding>)encoder videoFrame:(nullable VideoFrame *)frame {
    if (self.uploading){
        if(frame.isKeyFrame && self.hasCaptureAudio) self.hasKeyFrameVideo = YES;
        if(self.AVAlignment) [self pushSendBuffer:frame];
    }
}

#pragma mark -- StreamTcpSocketDelegate
- (void)socketStatus:(nullable id<StreamSocket>)socket status:(LiveState)status {
    if (status == LiveStart) {
        if (!self.uploading) {
            self.AVAlignment = NO;
            self.hasCaptureAudio = NO;
            self.hasKeyFrameVideo = NO;
            self.relativeTimestamps = 0;
            self.uploading = YES;
        }
    } else if(status == LiveStop || status == LiveError){
        self.uploading = NO;
    }
    dispatch_async(dispatch_get_main_queue(), ^{
        self.state = status;
        if (self.delegate && [self.delegate respondsToSelector:@selector(liveSession:liveStateDidChange:)]) {
            [self.delegate liveSession:self liveStateDidChange:status];
        }
    });
}

- (void)socketDidError:(nullable id<StreamSocket>)socket errorCode:(LiveSocketErrorCode)errorCode {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate && [self.delegate respondsToSelector:@selector(liveSession:errorCode:)]) {
            [self.delegate liveSession:self errorCode:errorCode];
        }
    });
}

- (void)socketDebug:(nullable id<StreamSocket>)socket debugInfo:(nullable LiveDebug *)debugInfo {
    self.debugInfo = debugInfo;
    if (self.showDebugInfo) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (self.delegate && [self.delegate respondsToSelector:@selector(liveSession:debugInfo:)]) {
                [self.delegate liveSession:self debugInfo:debugInfo];
            }
        });
    }
}

- (void)socketBufferStatus:(nullable id<StreamSocket>)socket status:(LiveBuffferState)status {
    if((self.captureType & LiveCaptureMaskVideo || self.captureType & LiveInputMaskVideo) && self.adaptiveBitrate){
        NSUInteger videoBitRate = [self.videoEncoder videoBitRate];
        if (status == LiveBuffferDecline) {
            if (videoBitRate < _videoConfiguration.videoMaxBitRate) {
                videoBitRate = videoBitRate + 50 * 1000;
                [self.videoEncoder setVideoBitRate:videoBitRate];
                NSLog(@"Increase bitrate %@", @(videoBitRate));
            }
        } else {
            if (videoBitRate > self.videoConfiguration.videoMinBitRate) {
                videoBitRate = videoBitRate - 100 * 1000;
                [self.videoEncoder setVideoBitRate:videoBitRate];
                NSLog(@"Decline bitrate %@", @(videoBitRate));
            }
        }
    }
}

#pragma mark -- Getter Setter
- (void)setRunning:(BOOL)running {
    if (_running == running) return;
    [self willChangeValueForKey:@"running"];
    _running = running;
    [self didChangeValueForKey:@"running"];
    [self.videoCaptureSource begin];
    self.audioCaptureSource.running = _running;
}

- (void)setPreviewImageView:(UIView *)previewImageView {
    [self willChangeValueForKey:@"previewImageView"];
    [self.videoCaptureSource setPreviewImageView:previewImageView];
    [self didChangeValueForKey:@"previewImageView"];
}

- (UIView *)previewImageView {
    return self.videoCaptureSource.previewImageView;
}

- (void)setCaptureDevicePosition:(AVCaptureDevicePosition)captureDevicePosition {
    [self willChangeValueForKey:@"captureDevicePosition"];
    [self.videoCaptureSource setCaptureDevicePosition:captureDevicePosition];
    [self didChangeValueForKey:@"captureDevicePosition"];
}

- (AVCaptureDevicePosition)captureDevicePosition {
    return self.videoCaptureSource.captureDevicePosition;
}

- (void)setBeautyFace:(BOOL)beautyFace {
    [self willChangeValueForKey:@"beautyFace"];
    [self.videoCaptureSource setBeautyFace:beautyFace];
    [self didChangeValueForKey:@"beautyFace"];
}

- (BOOL)saveLocalVideo{
    return self.videoCaptureSource.saveLocalVideo;
}

- (void)setSaveLocalVideo:(BOOL)saveLocalVideo{
    [self.videoCaptureSource setSaveLocalVideo:saveLocalVideo];
}


- (NSURL*)saveLocalVideoPath{
    return self.videoCaptureSource.saveLocalVideoPath;
}

- (void)setSaveLocalVideoPath:(NSURL*)saveLocalVideoPath{
   [self.videoCaptureSource setSaveLocalVideoPath:saveLocalVideoPath];
}

- (BOOL)beautyFace {
    return self.videoCaptureSource.beautyFace;
}

- (void)setBeautyLevel:(CGFloat)beautyLevel {
    [self willChangeValueForKey:@"beautyLevel"];
    [self.videoCaptureSource setBeautyLevel:beautyLevel];
    [self didChangeValueForKey:@"beautyLevel"];
}

- (CGFloat)beautyLevel {
    return self.videoCaptureSource.beautyLevel;
}

- (void)setBrightLevel:(CGFloat)brightLevel {
    [self willChangeValueForKey:@"brightLevel"];
    [self.videoCaptureSource setBrightLevel:brightLevel];
    [self didChangeValueForKey:@"brightLevel"];
}

- (CGFloat)brightLevel {
    return self.videoCaptureSource.brightLevel;
}

- (void)setZoomScale:(CGFloat)zoomScale {
    [self willChangeValueForKey:@"zoomScale"];
    [self.videoCaptureSource setZoomScale:zoomScale];
    [self didChangeValueForKey:@"zoomScale"];
}

- (CGFloat)zoomScale {
    return self.videoCaptureSource.zoomScale;
}

- (void)setTorch:(BOOL)torch {
    [self willChangeValueForKey:@"torch"];
    [self.videoCaptureSource setTorch:torch];
    [self didChangeValueForKey:@"torch"];
}

- (BOOL)torch {
    return self.videoCaptureSource.torch;
}

- (void)setMirror:(BOOL)mirror {
    [self willChangeValueForKey:@"mirror"];
    [self.videoCaptureSource setMirror:mirror];
    [self didChangeValueForKey:@"mirror"];
}

- (BOOL)mirror {
    return self.videoCaptureSource.mirror;
}

- (void)setMuted:(BOOL)muted {
    [self willChangeValueForKey:@"muted"];
    [self.audioCaptureSource setMuted:muted];
    [self didChangeValueForKey:@"muted"];
}

- (BOOL)muted {
    return self.audioCaptureSource.muted;
}

- (nullable UIImage *)currentImage{
    return self.videoCaptureSource.currentImage;
}

- (AudioCapture *)audioCaptureSource {
    if (!_audioCaptureSource) {
        if(self.captureType & LiveCaptureMaskAudio){
            _audioCaptureSource = [[AudioCapture alloc] initWithAudioConfiguration:_audioConfiguration];
            _audioCaptureSource.delegate = self;
        }
    }
    return _audioCaptureSource;
}

- (LiveVideoCapture *)videoCaptureSource {
    if (!_videoCaptureSource) {
        if(self.captureType & LiveCaptureMaskVideo){
            _videoCaptureSource = [[LiveVideoCapture alloc] initWith:_videoConfiguration];
            _videoCaptureSource.delegate = self;
        }
    }
    return _videoCaptureSource;
}

- (id<AudioEncoding>)audioEncoder {
    if (!_audioEncoder) {
        _audioEncoder = [[LFHardwareAudioEncoder alloc] initWithAudioStreamConfiguration:_audioConfiguration];
        [_audioEncoder setDelegate:self];
    }
    return _audioEncoder;
}

- (id<VideoEncoding>)videoEncoder {
    if (!_videoEncoder) {
        if([[UIDevice currentDevice].systemVersion floatValue] < 8.0){
            _videoEncoder = [[LFH264VideoEncoder alloc] initWithVideoStreamConfiguration:_videoConfiguration];
        }else{
            _videoEncoder = [[LFHardwareVideoEncoder alloc] initWithVideoStreamConfiguration:_videoConfiguration];
        }
        [_videoEncoder setDelegate:self];
    }
    return _videoEncoder;
}

- (id<StreamSocket>)socket {
    if (!_socket) {
        _socket = [[StreamRTMPSocket alloc] initWithStream:self.streamInfo reconnectInterval:self.reconnectInterval reconnectCount:self.reconnectCount];
        [_socket setDelegate:self];
    }
    return _socket;
}

- (LiveStreamInfo *)streamInfo {
    if (!_streamInfo) {
        _streamInfo = [[LiveStreamInfo alloc] init];
    }
    return _streamInfo;
}

- (dispatch_semaphore_t)lock{
    if(!_lock){
        _lock = dispatch_semaphore_create(1);
    }
    return _lock;
}

- (uint64_t)uploadTimestamp:(uint64_t)captureTimestamp{
    dispatch_semaphore_wait(self.lock, DISPATCH_TIME_FOREVER);
    uint64_t currentts = 0;
    currentts = captureTimestamp - self.relativeTimestamps;
    dispatch_semaphore_signal(self.lock);
    return currentts;
}

- (BOOL)AVAlignment{
    if((self.captureType & LiveCaptureMaskAudio || self.captureType & LiveInputMaskAudio) &&
       (self.captureType & LiveCaptureMaskVideo || self.captureType & LiveInputMaskVideo)
       ){
        if(self.hasCaptureAudio && self.hasKeyFrameVideo) return YES;
        else  return NO;
    }else{
        return YES;
    }
}

@end

File: LiveKit/LiveSession.swift
//
//  LiveSession.swift
//  LiveKit
//
//  Created by Davis, Tyler on 5/20/19.
//  Copyright © 2019 admin. All rights reserved.
//

import Foundation
import UIKit

/// Capture audio, video.
public enum LiveCaptureType: Int {
    case captureAudio
    case captureVideo
}

/// Mask for defining audio/video capture
public struct LiveCaptureTypeMask: OptionSet {
    public let rawValue: Int
    
    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    public static let captureAudio = LiveCaptureTypeMask(rawValue: 1 << LiveCaptureType.captureAudio.rawValue)
    public static let captureVideo = LiveCaptureTypeMask(rawValue: 1 << LiveCaptureType.captureVideo.rawValue)

    public static let captureAll: LiveCaptureTypeMask = [.captureAudio, .captureVideo]

}

public protocol LiveSessionDelegate: class {
    func stateChange(to state: LiveState)
    func liveDebugInfo(info: LiveDebug)
    func didError(with code: LiveSocketErrorCode)
}

public class LiveSession: NSObject {

    public weak var delegate: LiveSessionDelegate?

    /// View required for video display.
    public var previewImageView: UIView? {
        didSet {
            videoCaptureSource.previewImageView = previewImageView
        }
    }

    var isRunning = false
    var isUploading = false

    let audioConfiguration: LiveAudioConfiguration
    let videoConfiguration: LiveVideoConfiguration

    /// The stream control upload and package
    var streamInfo: LiveStreamInfo?

    /// The status of the stream
    var state: LiveState = .ready

    /// The captureType control inner or outer audio and video.
    let captureType: LiveCaptureTypeMask

    /// The reconnectInterval control reconnect timeInterval.
    var reconnectInterval: Int = 5

    /// The reconnectCount control reconnect count.
    var reconnectCount = 0

    /// The currentImage is videoCapture shot.
    var currentImage: UIImage? {
        return videoCaptureSource.currentImage
    }

    var lock = DispatchSemaphore(value: 1)
    var relativeTimestamp: UInt64 = 0
    var AVAlignment: Bool {
        guard captureType.contains(.captureAudio) &&
                captureType.contains(.captureVideo) else {
            return true
        }
        return hasCaptureAudio && hasKeyFrameVideo
    }
    
    var hasCaptureAudio = false
    var hasKeyFrameVideo = false

    lazy var socket: StreamRTMPSocket? = {
        let rtmpSocket = StreamRTMPSocket(stream: streamInfo, reconnectInterval: reconnectInterval, reconnectCount: reconnectCount)
        rtmpSocket?.setDelegate(self)
        return rtmpSocket
    }()
    
    lazy var videoCaptureSource: LiveVideoCapture = {
        let videoSource = LiveVideoCapture(with: videoConfiguration)
        videoSource.delegate = self
        return videoSource
    }()

    lazy var audioCaptureSource: AudioCapture? = {
        let audioSource = AudioCapture(audioConfiguration: audioConfiguration)
        audioSource?.delegate = self
        return audioSource
    }()

    lazy var videoEncoder: LFHardwareVideoEncoder? = {
        let encoder = LFHardwareVideoEncoder(videoStreamConfiguration: videoConfiguration)
        encoder?.setDelegate(self)
        return encoder
    }()

    lazy var audioEncoder: LFHardwareAudioEncoder? = {
        let encoder = LFHardwareAudioEncoder(audioStreamConfiguration: audioConfiguration)
        encoder?.setDelegate(self)
        return encoder
    }()
    
    public init(with audioConfiguration: LiveAudioConfiguration, videoConfiguration: LiveVideoConfiguration, captureType: LiveCaptureTypeMask = LiveCaptureTypeMask.captureAll) {
        self.audioConfiguration = audioConfiguration
        self.videoConfiguration = videoConfiguration
        self.captureType = captureType
    }

    deinit {
        videoCaptureSource.end()
    }

    /// Begins the session.
    /// Video is captured on the previewImageView.
    /// Doesn't upload to a stream.
    public func beginRunning() {
        guard !isRunning else {
            return
        }
        // run
        videoCaptureSource.begin()
        audioCaptureSource?.running = true
        isRunning = true
    }

    /// Ends the session.
    public func endRunning() {
        // end
        videoCaptureSource.end()
        audioCaptureSource?.running = false
        isRunning = false
    }

    /// Starts the stream.
    /// begins uploading when the socket connects to the server.
    ///
    /// - parameter streamInfo: LiveStreamInfo with the rtmp endpoint
    public func startLive(with streamInfo: LiveStreamInfo) {
        self.streamInfo = streamInfo
        streamInfo.videoConfiguration = videoConfiguration
        streamInfo.audioConfiguration = audioConfiguration
        socket?.start()
    }

    /// Stops uploading to the stream.
    public func stopLive() {
        isUploading = false
        socket?.stop()
        socket = nil
    }

    /// Set the camera zoom.
    /// Default is 1.0
    ///
    /// - parameter zoom: zoom will only work between 1.0 ~ 3.0
    public func setZoomScale(to zoom: CGFloat) {
        videoCaptureSource.zoomScale = zoom
    }

    /// Toggles the camera flash on/off.
    /// Default is off.
    public func toggleTorch() {
        let isOn = videoCaptureSource.torch
        videoCaptureSource.torch = !isOn
    }

    /// Toggle the camera position.
    /// Default is back.
    ///
    /// - parameter position: front or back
    public func toggleDevicePosition(to position: AVCaptureDevice.Position) {
        videoCaptureSource.captureDevicePosition = position
    }

    /// Send or mute audio.
    ///
    /// - parameter mute: default is on.
    public func muteAudio(_ mute: Bool) {
        audioCaptureSource?.muted = mute
    }

}

extension LiveSession {

    func uploadTimestamp(_ captureTimestamp: UInt64) -> UInt64 {
        let current: UInt64
        lock.wait()
        current = captureTimestamp - relativeTimestamp
        lock.signal()
        return current
    }

    func pushSendBuffer(_ frame: LFFrame) {
        if relativeTimestamp == 0 {
            relativeTimestamp = frame.timestamp
        }
        frame.timestamp = uploadTimestamp(frame.timestamp)
        socket?.send(frame)
    }

}

extension LiveSession: VideoCaptureDelegate {
    public func captureOutput(capture: LiveVideoCapture?, pixelBuffer: CVPixelBuffer?) {
        if isUploading {
            videoEncoder?.encodeVideoData(pixelBuffer, timeStamp: UInt64(CACurrentMediaTime()*1000))
        }
    }
}

extension LiveSession: AudioCaptureDelegate {
    public func captureOutput(_ capture: AudioCapture?, audioData: Data?) {
        if isUploading {
            audioEncoder?.encodeAudioData(audioData, timeStamp: UInt64(CACurrentMediaTime()*1000))
        }
    }
}

extension LiveSession: VideoEncodingDelegate {
    public func videoEncoder(_ encoder: VideoEncoding?, videoFrame frame: VideoFrame?) {
        guard let frame = frame, isUploading else {
            return
        }
        if frame.isKeyFrame && hasCaptureAudio {
            hasKeyFrameVideo = true
        }
        if AVAlignment {
            pushSendBuffer(frame)
        }
    }
}

extension LiveSession: AudioEncodingDelegate {
    public func audioEncoder(_ encoder: AudioEncoding?, audioFrame frame: AudioFrame?) {
        guard let frame = frame, isUploading else {
            return
        }
        hasCaptureAudio = true
        if AVAlignment {
            pushSendBuffer(frame)
        }
    }
}

extension LiveSession: StreamSocketDelegate {

    public func socketBufferStatus(_ socket: StreamSocket?, status: LiveBufferState) {
        // TODO (WTD)
//        if((self.captureType & LiveCaptureMaskVideo || self.captureType & LiveInputMaskVideo) && self.adaptiveBitrate){
//            NSUInteger videoBitRate = [self.videoEncoder videoBitRate];
//            if (status == LiveBufferDecline) {
//                if (videoBitRate < _videoConfiguration.videoMaxBitRate) {
//                    videoBitRate = videoBitRate + 50 * 1000;
//                    [self.videoEncoder setVideoBitRate:videoBitRate];
//                    NSLog(@"Increase bitrate %@", @(videoBitRate));
//                }
//            } else {
//                if (videoBitRate > self.videoConfiguration.videoMinBitRate) {
//                    videoBitRate = videoBitRate - 100 * 1000;
//                    [self.videoEncoder setVideoBitRate:videoBitRate];
//                    NSLog(@"Decline bitrate %@", @(videoBitRate));
//                }
//            }
//        }
//
    }
    
    public func socketStatus(_ socket: StreamSocket?, status: LiveState) {
        switch status {
        case .broadcasting:
            hasCaptureAudio = false
            hasKeyFrameVideo = false
            relativeTimestamp = 0
            isUploading = true
        case .ended, .error:
            isUploading = false
        default:
            break
        }

        DispatchQueue.main.async { [weak self] in
            self?.state = status
            self?.delegate?.stateChange(to: status)
        }
    }
    
    public func socketDidError(_ socket: StreamSocket?, errorCode: LiveSocketErrorCode) {
        DispatchQueue.main.async { [weak self] in
            self?.delegate?.didError(with: errorCode)
        }
    }

}

// TODO: Add funciontality
///** The mirror control mirror of front camera is on or off */
//@property (nonatomic, assign) BOOL mirror;
//
///** The muted control callbackAudioData,muted will memset 0.*/
//@property (nonatomic, assign) BOOL muted;
//
///*  The adaptiveBitrate control auto adjust bitrate. Default is NO */
//@property (nonatomic, assign) BOOL adaptiveBitrate;
///** The showDebugInfo control streamInfo and uploadInfo(1s) *.*/
//@property (nonatomic, assign) BOOL showDebugInfo;
//
///* The saveLocalVideo is save the local video */
//@property (nonatomic, assign) BOOL saveLocalVideo;
//
///* The saveLocalVideoPath is save the local video  path */
//@property (nonatomic, strong, nullable) NSURL *saveLocalVideoPath;
//
